{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "#os.environ[\"CUDA_DEVICE_ORDER\"]=\"PCI_BUS_ID\"   # see issue #152\n",
    "#os.environ[\"CUDA_VISIBLE_DEVICES\"]=\"0\"\n",
    "import torch\n",
    "GPU = torch.cuda.is_available()\n",
    "print(GPU)\n",
    "from torch.cuda.amp import autocast, GradScaler\n",
    "scaler = GradScaler()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GPU = False\n",
    "METAPRECISION = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-26 09:53:11.139591: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----vocab size 129407 ------\n",
      "MODEL SIZE: \n",
      "1280\n",
      "model initialized\n",
      "Found PyTorch checkpoint at  ckpt/pretrain_progen_full.pth\n",
      "Loading instead of converting from TensorFlow\n",
      "previous checkpoint loaded\n",
      "previous checkpoint loaded in GPU\n"
     ]
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "from __future__ import division\n",
    "import os\n",
    "import sys\n",
    "import torch\n",
    "import tqdm\n",
    "import pdb\n",
    "import numpy as np\n",
    "import platform\n",
    "import hashlib\n",
    "import pytorch_transformer\n",
    "import re\n",
    "import argparse\n",
    "import tensorflow as tf\n",
    "from tensorflow.python import pywrap_tensorflow\n",
    "import torch.nn.functional as F\n",
    "#from torch.utils.tensorboard import SummaryWriter\n",
    "from transformProtein import transformProtein\n",
    "from ProteinDataset_uid import ProteinDataset\n",
    "from torch.utils.data import Dataset, DataLoader\n",
    "import pickle\n",
    "import time\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "torch.cuda.empty_cache()\n",
    "load_model_path = 'ckpt/' # just the folder itself\n",
    "\n",
    "seq_length = 511\n",
    "embedding_dim = 1280\n",
    "num_layers = 36\n",
    "vocab_loc = 'mapping_files/vocab.txt'\n",
    "\n",
    "use_py3 = platform.python_version()[0] == '3'\n",
    "vocab = open(vocab_loc).readlines() if not use_py3 else open(vocab_loc, encoding='utf-8').read().split('\\n')[:-1]\n",
    "vocab = list(map(lambda x: x.split(' ')[0], vocab))\n",
    "vocab_size = len(vocab)\n",
    "print('-----vocab size',vocab_size,'------')\n",
    "\n",
    "class TiedEmbeddingSoftmax(torch.nn.Module):\n",
    "\n",
    "    def __init__(self, vocab_size=vocab_size, embedding_size=embedding_dim, **kwargs):\n",
    "        super(TiedEmbeddingSoftmax, self).__init__()\n",
    "        self.w = torch.nn.Parameter(torch.normal(0., 1e-2, size=(vocab_size, embedding_size)))\n",
    "        self.b = torch.nn.Parameter(torch.zeros(vocab_size))\n",
    "\n",
    "    def forward(self, inputs, embed=True):\n",
    "        with autocast(enabled=METAPRECISION):\n",
    "            if embed:\n",
    "                return torch.nn.functional.embedding(inputs, self.w)\n",
    "            else:\n",
    "                return torch.tensordot(inputs, self.w.t(), 1) + self.b\n",
    "\n",
    "class CTRLmodel(torch.nn.Module):\n",
    "    def __init__(self):\n",
    "        super(CTRLmodel,self).__init__()\n",
    "        self.tied_embedding_softmax = TiedEmbeddingSoftmax()\n",
    "        self.encoder = pytorch_transformer.Encoder()\n",
    "\n",
    "    def forward(self, inputs):\n",
    "        with autocast(enabled=METAPRECISION):\n",
    "            x = self.tied_embedding_softmax(inputs, embed=True)\n",
    "            x = self.encoder(x)\n",
    "            x = self.tied_embedding_softmax(x, embed=False)\n",
    "        return x\n",
    "\n",
    "    def loadCheckpoint(self, model_path, num_layers):\n",
    "        if os.path.exists(model_path):\n",
    "            print('Found PyTorch checkpoint at ', model_path)\n",
    "            print('Loading instead of converting from TensorFlow')\n",
    "            checkpoint = torch.load(model_path, map_location=torch.device('cpu'))\n",
    "            self.tied_embedding_softmax.load_state_dict({\n",
    "                'w': checkpoint.pop('tied_embedding_softmax.w', None),\n",
    "                'b': checkpoint.pop('tied_embedding_softmax.b', None)\n",
    "            })\n",
    "            self.encoder.load_state_dict({key.replace(\"encoder.\", \"\"): value for key, value in checkpoint.items()})\n",
    "        else:\n",
    "            print('Could not find PyTorch checkpoint')\n",
    "            sys.exit()\n",
    "\n",
    "model = CTRLmodel()\n",
    "print('model initialized')\n",
    "\n",
    "curr_model_path = load_model_path+'pretrain_progen_full.pth'\n",
    "reader = model.loadCheckpoint(model_path=curr_model_path, num_layers = num_layers)\n",
    "print('previous checkpoint loaded')\n",
    "            \n",
    "if METAPRECISION:\n",
    "    model = model.to(torch.float16)\n",
    "if GPU:\n",
    "    model = model.cuda()\n",
    "    print('previous checkpoint loaded in GPU')\n",
    "\n",
    "optimizer = torch.optim.Adam(model.parameters()) #lr, betas\n",
    "\n",
    "model.eval()\n",
    "\n",
    "with open(os.path.join('mapping_files/','taxa_to_lineage.p'),'rb') as handle:\n",
    "    taxa_to_lineage = pickle.load(handle)\n",
    "with open('mapping_files/taxa_to_ctrl_idx.p','rb') as handle:\n",
    "    taxa_to_ctrl_idx = pickle.load(handle)\n",
    "with open('mapping_files/kw_to_ctrl_idx.p','rb') as handle:\n",
    "    kw_to_ctrl_idx = pickle.load(handle)\n",
    "with open('mapping_files/aa_to_ctrl_idx.p','rb') as handle:\n",
    "    aa_to_ctrl_idx = pickle.load(handle)\n",
    "with open('mapping_files/kw_to_name.p2','rb') as handle:\n",
    "    kw_to_name = pickle.load(handle)\n",
    "#with open('mapping_files/taxid_to_name.p2','rb') as handle:\n",
    "#    taxid_to_name = pickle.load(handle)\n",
    "# MANCANO:\n",
    "# kw_to_lineage.p\n",
    "# kw_to_parents.p\n",
    "# taxa_to_parents.p\n",
    "    \n",
    "def flipdict(my_map):\n",
    "    return {v: k for k, v in my_map.items()}\n",
    "ctrl_idx_to_aa = flipdict(aa_to_ctrl_idx)\n",
    "ctrl_idx_to_kw = flipdict(kw_to_ctrl_idx)\n",
    "ctrl_idx_to_taxa = flipdict(taxa_to_ctrl_idx)\n",
    "name_to_kw = flipdict(kw_to_name)\n",
    "\n",
    "def predict_fn(inputs):\n",
    "    with torch.no_grad():\n",
    "        inputs = torch.tensor(inputs)\n",
    "        if GPU:\n",
    "            inputs = inputs.cuda()            \n",
    "        output = model(inputs)\n",
    "        output = output[:,:,-26:-1] # remove non-AA token logits\n",
    "        return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Random Selection:\n",
      "KRTRFRKQHRGR\n",
      "2000\n",
      "{'entryType': 'UniProtKB unreviewed (TrEMBL)', 'primaryAccession': 'Q8MEP1', 'uniProtkbId': 'Q8MEP1_9ROSI', 'entryAudit': {'firstPublicDate': '2002-10-01', 'lastAnnotationUpdateDate': '2019-12-11', 'lastSequenceUpdateDate': '2002-10-01', 'entryVersion': 25, 'sequenceVersion': 1}, 'annotationScore': 1.0, 'organism': {'scientificName': 'Hibiscus pentaphyllus', 'taxonId': 183255, 'evidences': [{'evidenceCode': 'ECO:0000313', 'source': 'EMBL', 'id': 'AAM50369.1'}], 'lineage': ['Eukaryota', 'Viridiplantae', 'Streptophyta', 'Embryophyta', 'Tracheophyta', 'Spermatophyta', 'Magnoliopsida', 'eudicotyledons', 'Gunneridae', 'Pentapetalae', 'rosids', 'malvids', 'Malvales', 'Malvaceae', 'Malvoideae', 'Hibiscus']}, 'proteinExistence': '4: Predicted', 'proteinDescription': {'submissionNames': [{'fullName': {'evidences': [{'evidenceCode': 'ECO:0000313', 'source': 'EMBL', 'id': 'AAM50369.1'}], 'value': 'Ribosomal protein 16'}}], 'flag': 'Fragment'}, 'genes': [{'geneName': {'evidences': [{'evidenceCode': 'ECO:0000313', 'source': 'EMBL', 'id': 'AAM50369.1'}], 'value': 'rpl16'}}], 'features': [{'type': 'Non-terminal residue', 'location': {'start': {'value': 1, 'modifier': 'EXACT'}, 'end': {'value': 1, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0000313', 'source': 'EMBL', 'id': 'AAM50369.1'}]}, {'type': 'Non-terminal residue', 'location': {'start': {'value': 12, 'modifier': 'EXACT'}, 'end': {'value': 12, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0000313', 'source': 'EMBL', 'id': 'AAM50369.1'}]}], 'geneLocations': [{'evidences': [{'evidenceCode': 'ECO:0000313', 'source': 'EMBL', 'id': 'AAM50369.1'}], 'geneEncodingType': 'Chloroplast'}], 'keywords': [{'evidences': [{'evidenceCode': 'ECO:0000313', 'source': 'EMBL', 'id': 'AAM50369.1'}], 'id': 'KW-0150', 'category': 'Cellular component', 'name': 'Chloroplast'}, {'evidences': [{'evidenceCode': 'ECO:0000313', 'source': 'EMBL', 'id': 'AAM50369.1'}], 'id': 'KW-0934', 'category': 'Cellular component', 'name': 'Plastid'}, {'evidences': [{'evidenceCode': 'ECO:0000313', 'source': 'EMBL', 'id': 'AAM50369.1'}], 'id': 'KW-0687', 'category': 'Molecular function', 'name': 'Ribonucleoprotein'}, {'evidences': [{'evidenceCode': 'ECO:0000313', 'source': 'EMBL', 'id': 'AAM50369.1'}], 'id': 'KW-0689', 'category': 'Molecular function', 'name': 'Ribosomal protein'}], 'references': [{'citation': {'id': 'IND23289778', 'citationType': 'journal article', 'authors': ['Pfeil B.E.', 'Brubaker C.L.', 'Craven L.A.', 'Crisp M.D.'], 'citationCrossReferences': [{'database': 'AGRICOLA', 'id': 'IND23289778'}], 'title': 'Phylogeny of Hibiscus and the tribe Hibisceae (Malvaceae) using chloroplast DNA sequences of ndhF and the rpl16 intron.', 'publicationDate': '2002', 'journal': 'Syst. Bot.', 'firstPage': '333', 'lastPage': '350', 'volume': '27'}, 'referencePositions': ['NUCLEOTIDE SEQUENCE'], 'evidences': [{'evidenceCode': 'ECO:0000313', 'source': 'EMBL', 'id': 'AAM50369.1'}]}], 'uniProtKBCrossReferences': [{'database': 'EMBL', 'id': 'AF384597', 'properties': [{'key': 'ProteinId', 'value': 'AAM50369.1'}, {'key': 'Status', 'value': '-'}, {'key': 'MoleculeType', 'value': 'Genomic_DNA'}]}, {'database': 'GO', 'id': 'GO:0009507', 'properties': [{'key': 'GoTerm', 'value': 'C:chloroplast'}, {'key': 'GoEvidenceType', 'value': 'IEA:UniProtKB-KW'}]}, {'database': 'GO', 'id': 'GO:0005840', 'properties': [{'key': 'GoTerm', 'value': 'C:ribosome'}, {'key': 'GoEvidenceType', 'value': 'IEA:UniProtKB-KW'}]}], 'sequence': {'value': 'KRTRFRKQHRGR', 'length': 12, 'molWeight': 1626, 'crc64': 'DDD227C351D32409', 'md5': 'AD25132B3F22242020C766A31BB2CDF9'}, 'extraAttributes': {'countByFeatureType': {'Non-terminal residue': 2}, 'uniParcId': 'UPI000008BBF1'}}\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "import os\n",
    "# paths to the saved .p files\n",
    "random_selection_file = os.path.join(\"data_random_selection_uniParc\", \"random_selection_2000.p\")\n",
    "#data_file = os.path.join(\"data\", \"filtered_data_\" + query + \".p\")\n",
    "\n",
    "# Reload random_selection from the .p file\n",
    "random_selection = False\n",
    "with open(random_selection_file, \"rb\") as file:\n",
    "    random_selection = pickle.load(file)\n",
    "\n",
    "# Reload filtered_data from the .p file\n",
    "#filtered_data = False\n",
    "#with open(data_file, \"rb\") as file:\n",
    "#    filtered_data = pickle.load(file)\n",
    "\n",
    "# Print or use the reloaded datasets as needed\n",
    "print(\"Random Selection:\")\n",
    "for entry in random_selection:\n",
    "    print(entry[\"sequence\"][\"value\"])\n",
    "    print(len(random_selection))\n",
    "    print(entry)\n",
    "    break\n",
    "\n",
    "#print(\"---\")\n",
    "\n",
    "#print(\"Filtered Data:\")\n",
    "#for entry in filtered_data:\n",
    "#    print(entry[\"sequence\"])'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tax lineage problems: 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mnicolini/miniconda3/envs/vPro/lib/python3.11/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 93032 was translated into 58707\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    }
   ],
   "source": [
    "import ete3\n",
    "ncbi = ete3.NCBITaxa()\n",
    "count = 0\n",
    "inner_count = 0\n",
    "for entry in random_selection:\n",
    "    taxid = int(entry['organism']['taxonId']) # taxonomy id from NCBI\n",
    "    inner_count = 0\n",
    "    #print(taxid)\n",
    "    try:\n",
    "        tax_lineage1 = ncbi.get_lineage(taxid)\n",
    "        # remove root taxonomy id\n",
    "        tax_lineage1 = [element for element in tax_lineage1 if element not in [1, 2, 131567, 10239]]\n",
    "        #tax_lineage2 = taxa_to_lineage[taxid] # make lineage in ncbi ids\n",
    "        #print(tax_lineage1)\n",
    "        res = []\n",
    "        for ite in tax_lineage1:\n",
    "            try:\n",
    "                tmp = taxa_to_ctrl_idx[ite]\n",
    "                res.append(tmp)\n",
    "            except:\n",
    "                inner_count += 1\n",
    "        if res == [] and inner_count >= 1:\n",
    "            count += 1\n",
    "    except:\n",
    "        count += 1\n",
    "        print(\"FATAL taxa conversion\")\n",
    "        #tax_lineage = [taxa_to_ctrl_idx[ite] for ite in tax_lineage1] # now translated as ctrl code indices\n",
    "print(\"Tax lineage problems:\", count)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p>To evaluate the generated sequence using teacher forcing:\n",
    "\n",
    "1. Generate the Sequence: The transformer model with teacher forcing generates the complete sequence. \n",
    "At each time step, we provide the true input sequence (+ 1 versus previous step), as input to the model.\n",
    "\n",
    "2. Calculate Metrics: Accuracy, soft accuracy (based on BLOSUM matrix), perplexity. \n",
    "\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def teacher_forcing_generation(input_sequence, tax_lineage, penalty, topk):\n",
    "    key_len = 0\n",
    "    res = \"\"\n",
    "    tokens_prob = []\n",
    "    true_tokens_index_in_prob = []\n",
    "    key_len = len(tax_lineage) # len(kw_lineage+tax_lineage)\n",
    "    for i in range(1, len(input_sequence)):\n",
    "        iteration_input_prefix = input_sequence[:i]\n",
    "        seed_seq = [aa_to_ctrl_idx[ii] for ii in iteration_input_prefix]\n",
    "        generate_num = key_len + len(seed_seq) + 1 # how many tokens to generate, here only one\n",
    "        seq_length = min(generate_num, 511)\n",
    "        text = tax_lineage + seed_seq # tax_lineage + kw_lineage + seed_seq\n",
    "        padded_text = text + [0] * (generate_num - len(text))\n",
    "        tokens_generated = np.tile(padded_text, (1,1))\n",
    "        for token in range(len(text)-1, generate_num-1):\n",
    "            prompt_logits = predict_fn(tokens_generated[:, :seq_length]).squeeze()\n",
    "            _token = token if token < seq_length else -1\n",
    "            prompt_logits = prompt_logits.cpu().detach().numpy()\n",
    "            #print(tokens_generated[:, :seq_length])\n",
    "            #print(prompt_logits)\n",
    "            if penalty>0:\n",
    "                penalized_so_far = set()\n",
    "                # variable token_flag for first amminoacids (to count them if they are less that 4)\n",
    "                if token >= key_len + 3:\n",
    "                    token_flag = 3  \n",
    "                elif token - key_len - 3 <= 0:\n",
    "                    token_flag = 0\n",
    "                else:\n",
    "                    token_flag = token\n",
    "                for _ in range(token-token_flag,token+1):\n",
    "                    generated_token = tokens_generated[0][_] - (vocab_size-26) # added\n",
    "                    if generated_token in penalized_so_far:\n",
    "                        continue\n",
    "                    penalized_so_far.add(generated_token)\n",
    "                    prompt_logits[_token][generated_token] /= penalty  \n",
    "            # compute probabilities from logits\n",
    "            prompt_probs = np.exp(prompt_logits[_token])\n",
    "            prompt_probs = prompt_probs / sum(prompt_probs)\n",
    "            pruned_list = np.argsort(prompt_probs)[::-1]\n",
    "            codice_aa_della_true_sequence = seed_seq[_token - key_len]\n",
    "            idx_true = codice_aa_della_true_sequence - vocab_size + 26\n",
    "            true_tokens_index_in_prob.append([idx_true])\n",
    "            tokens_prob.append([prompt_probs.tolist()])\n",
    "\n",
    "            if topk==1:\n",
    "                idx = pruned_list[0]\n",
    "            else:\n",
    "                pruned_list = pruned_list[:topk]\n",
    "                chosen_idx = torch.distributions.categorical.Categorical(logits=torch.tensor(np.expand_dims(prompt_logits[_token][pruned_list],0))).sample().numpy()[0]\n",
    "                idx = pruned_list[chosen_idx]\n",
    "            # assign the token for generation\n",
    "            idx += (vocab_size-26) # added to convert 0 AA to original ctrl idx\n",
    "            tokens_generated[0][token+1] = idx\n",
    "        tokens_generated_so_far = tokens_generated[0].squeeze()[:token+2]\n",
    "        tokens_generated_so_far = tokens_generated_so_far[(tokens_generated_so_far>=(vocab_size-26)) & (tokens_generated_so_far<(vocab_size-1))]\n",
    "        tokens_generated_so_far = ''.join([ctrl_idx_to_aa[c] for c in tokens_generated_so_far])\n",
    "        query = tokens_generated_so_far[len(seed_seq):]\n",
    "        res += query\n",
    "    return res, tokens_prob, true_tokens_index_in_prob\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FATAL keywords conversion for entry:  1552\n",
      "FATAL keywords conversion for entry:  91360\n",
      "FATAL keywords conversion for entry:  228952\n",
      "FATAL keywords conversion for entry:  13243\n",
      "FATAL keywords conversion for entry:  35327\n",
      "FATAL keywords conversion for entry:  37042\n",
      "FATAL keywords conversion for entry:  2424\n",
      "FATAL keywords conversion for entry:  195088\n",
      "FATAL keywords conversion for entry:  252230\n",
      "FATAL keywords conversion for entry:  3505\n",
      "FATAL keywords conversion for entry:  4615\n",
      "FATAL keywords conversion for entry:  11023\n",
      "FATAL keywords conversion for entry:  28689\n",
      "FATAL keywords conversion for entry:  38402\n",
      "FATAL keywords conversion for entry:  198940\n",
      "FATAL keywords conversion for entry:  1951\n",
      "FATAL keywords conversion for entry:  4615\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import pickle\n",
    "# print(\"on data: \", query)\n",
    "penalty = 0\n",
    "topk = 1\n",
    "predicted = []\n",
    "true_tokens_index_in_probs_all = []\n",
    "tokens_probs_all = []\n",
    "true_value = []\n",
    "for entry in random_selection:\n",
    "    input_seq = entry['sequence']['value']\n",
    "    taxid = int(entry['organism']['taxonId'])\n",
    "    try:\n",
    "        tax_lineage = ncbi.get_lineage(taxid)\n",
    "        # remove root taxonomy id\n",
    "        tax_lineage = [element for element in tax_lineage1 if element not in [1, 2, 131567, 10239]]\n",
    "        res = []\n",
    "        inner_count = 0\n",
    "        for ite in tax_lineage:\n",
    "            try:\n",
    "                tmp = taxa_to_ctrl_idx[ite]\n",
    "                res.append(tmp)\n",
    "            except:\n",
    "                inner_count += 1\n",
    "        if res == [] and inner_count >= 1:\n",
    "            print(\"ALERT: error in taxonmy conversions for entry: \", taxid)\n",
    "        tax_lineage = res\n",
    "    except:\n",
    "        print(\"FATAL taxa conversion for entry: \", taxid)\n",
    "    # Access keywords\n",
    "    try:\n",
    "        keywords = [keyword['name'] for keyword in entry['keywords']]\n",
    "        key_ctrl = []\n",
    "        for key in keywords:\n",
    "            ite = name_to_kw[key]\n",
    "            tmp = kw_to_ctrl_idx[ite]\n",
    "            key_ctrl.append(tmp)\n",
    "    except:\n",
    "        print(\"FATAL keywords conversion for entry: \", taxid)\n",
    "    # FOR TESTIG the code:\n",
    "    # input_seq = input_seq[0:4]\n",
    "    \n",
    "    tax_lineage = key_ctrl + tax_lineage\n",
    "\n",
    "    # if no tax keys:\n",
    "    # tax_lineage = []\n",
    "    \n",
    "    try:\n",
    "        res, tokens_prob, true_tokens_index_in_prob = teacher_forcing_generation(input_seq, tax_lineage,\n",
    "                                                                                 penalty, topk)\n",
    "    except:\n",
    "        print(\"ALERT: skipped for error the entry: \", entry)\n",
    "        continue\n",
    "    # print(\"Input: \", input_seq)\n",
    "    # print(\"Res: \", res)\n",
    "    true_value.append(input_seq)\n",
    "    true_tokens_index_in_probs_all.append(true_tokens_index_in_prob)\n",
    "    tokens_probs_all.append(tokens_prob)\n",
    "    predicted.append(input_seq[0] + res)\n",
    "\n",
    "    # FOR TESTIG:\n",
    "    #break\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a directory named \"ID_test_data\" in the current working directory if it doesn't exist\n",
    "data_dir = \"new_random_selection_2000_sequences_p_0_topk_1_kw_taxa_yes/\"\n",
    "os.makedirs(data_dir, exist_ok=True)\n",
    "\n",
    "# TODO, salvare tokens_prob e true_tokens_index_in_prob\n",
    "tokens_probs_data_file = os.path.join(data_dir, \"tokens_probs_data.p\")\n",
    "with open(tokens_probs_data_file, \"wb\") as file:\n",
    "    pickle.dump(tokens_probs_all, file)\n",
    "    \n",
    "true_tokens_index_data_file = os.path.join(data_dir, \"true_tokens_index_data.p\")\n",
    "with open(true_tokens_index_data_file, \"wb\") as file:\n",
    "    pickle.dump(true_tokens_index_in_probs_all, file)\n",
    "\n",
    "# Save ID_test_data as a .p file\n",
    "predicted_data_file = os.path.join(data_dir, \"predicted_data.p\")\n",
    "with open(predicted_data_file, \"wb\") as file:\n",
    "    pickle.dump(predicted, file)\n",
    "    \n",
    "# Save ID_test_data as a .p file\n",
    "true_data_file = os.path.join(data_dir, \"true_data.p\")\n",
    "with open(true_data_file, \"wb\") as file:\n",
    "    pickle.dump(true_value, file)\n",
    "torch.cuda.empty_cache()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(input_seq[0] + res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(input_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(input_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Evaluations</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Da implementare compute_perplexity\n",
    "from nltk import ngrams\n",
    "from sklearn.metrics import accuracy_score\n",
    "import math\n",
    "import blosum as bl\n",
    "import torch\n",
    "from torcheval.metrics.text import Perplexity\n",
    "import decimal\n",
    "import statistics\n",
    "# desired precision for printing float values\n",
    "decimal.getcontext().prec = 18\n",
    "#  BLOSUM62 is the matrix built using sequences with less than 62% similarity \n",
    "# (sequences with â‰¥ 62% identity were clustered)\n",
    "\n",
    "def compute_hard_accuracy(true_string, predicted_string):\n",
    "    # percentage of predictions that exactly match the ground truth labels\n",
    "    total_characters = len(true_string)\n",
    "    assert total_characters == len(predicted_string)\n",
    "    # Convert strings to lists of characters\n",
    "    true_list = list(true_string)\n",
    "    predicted_list = list(predicted_string)\n",
    "    # Compute hard accuracy using accuracy_score function\n",
    "    hard_accuracy = accuracy_score(true_list, predicted_list)\n",
    "    return hard_accuracy\n",
    "\n",
    "def compute_all_hard_accuracies(true_sequences, predicted_sequences):\n",
    "    num_seq = len(true_sequences)\n",
    "    assert num_seq == len(predicted_sequences)\n",
    "    accuracy_all = []\n",
    "    accuracy_dict = {}\n",
    "    for i in range(0, 8000, 50):\n",
    "        accuracy_dict[i] = []\n",
    "    for i in range(num_seq):\n",
    "        sequence = true_sequences[i]\n",
    "        predicted_sequence = predicted_sequences[i]\n",
    "        tmp_acc = compute_hard_accuracy(sequence, predicted_sequence)\n",
    "        accuracy_all.append(tmp_acc)\n",
    "        for j in range(0, len(sequence)-49, 50):\n",
    "            tmp_acc = compute_hard_accuracy(sequence[j:j+50], predicted_sequence[j:j+50])\n",
    "            accuracy_dict[j].append(float(tmp_acc))       \n",
    "    result = sum(accuracy_all) / num_seq\n",
    "    for key in accuracy_dict:\n",
    "        if len(accuracy_dict[key]) <= 0:\n",
    "            break\n",
    "        accuracy_dict[key] = sum(accuracy_dict[key])/len(accuracy_dict[key])\n",
    "    st_deviation = statistics.stdev(accuracy_all)\n",
    "    result = (result, st_deviation)\n",
    "    return result, accuracy_dict\n",
    "\n",
    "def compute_soft_accuracies(true_sequences, predicted_sequences, threshold = 0):\n",
    "    # soft accuracy takes into account partial matches or similarities \n",
    "    # between the predictions and the ground truth labels\n",
    "    bl_matrix = bl.BLOSUM(62)\n",
    "    num_seq = len(true_sequences)\n",
    "    assert num_seq == len(predicted_sequences)\n",
    "    accuracy_all = []\n",
    "    accuracy_dict = {}\n",
    "    for i in range(0, 8000, 50):\n",
    "        accuracy_dict[i] = []\n",
    "    for i in range(num_seq):\n",
    "        sequence = true_sequences[i]\n",
    "        predicted_sequence = predicted_sequences[i]\n",
    "        tmp_acc = calculate_soft_accuracy(sequence, predicted_sequence, bl_matrix, threshold)\n",
    "        accuracy_all.append(tmp_acc)\n",
    "        for j in range(0, len(sequence)-49, 50):\n",
    "            tmp_acc = calculate_soft_accuracy(sequence[j:j+50], predicted_sequence[j:j+50], bl_matrix, threshold)\n",
    "            accuracy_dict[j].append(tmp_acc) \n",
    "    result = sum(accuracy_all) / num_seq\n",
    "    for key in accuracy_dict:\n",
    "        if len(accuracy_dict[key]) <= 0:\n",
    "            break\n",
    "        accuracy_dict[key] = sum(accuracy_dict[key])/len(accuracy_dict[key])\n",
    "    st_deviation = statistics.stdev(accuracy_all)\n",
    "    result = (result, st_deviation)\n",
    "    return result, accuracy_dict\n",
    "\n",
    "def calculate_soft_accuracy(true_sequence, predicted_sequence, bl_matrix, threshold = 0):\n",
    "    total_characters = len(true_sequence)\n",
    "    assert total_characters == len(predicted_sequence)\n",
    "    partial_matches = 0\n",
    "    for true_aa, predicted_aa in zip(true_sequence, predicted_sequence):\n",
    "        similarity_score = bl_matrix[true_aa][predicted_aa]\n",
    "        if similarity_score >= threshold:\n",
    "            partial_matches += 1\n",
    "    soft_accuracy = partial_matches / total_characters\n",
    "    return soft_accuracy\n",
    "\n",
    "def compute_perplexity(probs, indexes):    \n",
    "    metric = Perplexity()\n",
    "    input = probs\n",
    "    target = indexes\n",
    "    metric.update(input, target)\n",
    "    perplexity = metric.compute()\n",
    "    return perplexity\n",
    "\n",
    "def compute_perplexities(probs_all, indexes_all):\n",
    "    num_seq = len(indexes_all)\n",
    "    try:\n",
    "        indexes_all[0][0][0]\n",
    "    except TypeError: \n",
    "        indexes = torch.tensor(indexes_all)\n",
    "        probs = torch.tensor(probs_all)\n",
    "        tmp_perplexity = compute_perplexity(probs, indexes)\n",
    "        return tmp_perplexity\n",
    "    assert num_seq == len(probs_all)\n",
    "    perplexity_all = []\n",
    "    perplexity_dict = {}\n",
    "    for i in range(0, 8000, 50):\n",
    "        perplexity_dict[i] = []\n",
    "    for i in range(num_seq):\n",
    "        indexes = torch.tensor(indexes_all[i])\n",
    "        probs = torch.tensor(probs_all[i])\n",
    "        tmp_perplexity = compute_perplexity(probs, indexes)\n",
    "        perplexity_all.append(tmp_perplexity)\n",
    "        for j in range(0, len(indexes)-49, 50):\n",
    "            tmp_acc = compute_perplexity(probs[j:j+50], indexes[j:j+50])\n",
    "            perplexity_dict[j].append(tmp_acc)  \n",
    "    result = sum(perplexity_all) / num_seq\n",
    "    for key in perplexity_dict:\n",
    "        if len(perplexity_dict[key]) <= 0:\n",
    "            break\n",
    "        perplexity_dict[key] = float(sum(perplexity_dict[key])/len(perplexity_dict[key]))\n",
    "    st_deviation = np.std(perplexity_all) # OR np.std\n",
    "    result = (float(result), float(st_deviation))\n",
    "    return result, perplexity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "# paths to the saved .p files\n",
    "data_dir = \"results_noGit/new_random_selection_100_sequences_p_1.2_topk_3_kw_taxa_no/\"\n",
    "predicted_data_file = os.path.join(data_dir, \"predicted_data.p\")\n",
    "true_data_file = os.path.join(data_dir, \"true_data.p\")\n",
    "indexes_all_data_file = os.path.join(data_dir, \"true_tokens_index_data.p\")\n",
    "probs_all_data_file = os.path.join(data_dir, \"tokens_probs_data.p\")\n",
    "\n",
    "# Reload predicted_data from the .p file\n",
    "predicted_data = False\n",
    "with open(predicted_data_file, \"rb\") as file:\n",
    "    predicted_data = pickle.load(file)\n",
    "\n",
    "# Reload true_data from the .p file\n",
    "true_data = False\n",
    "with open(true_data_file, \"rb\") as file:\n",
    "    true_data = pickle.load(file)\n",
    "    \n",
    "# Reload probs_all_data from the .p file\n",
    "probs_all_data = False\n",
    "with open(probs_all_data_file, \"rb\") as file:\n",
    "    probs_all_data = pickle.load(file)\n",
    "    \n",
    "# Reload indexes_all_data from the .p file\n",
    "indexes_all_data = False\n",
    "with open(indexes_all_data_file, \"rb\") as file:\n",
    "    indexes_all_data = pickle.load(file)\n",
    "\n",
    "# Compute hard accuracy\n",
    "accuracy, accuracy_dict = compute_all_hard_accuracies(true_data, predicted_data)\n",
    "\n",
    "# Compute perplexity\n",
    "perplexity, perplexity_dict = compute_perplexities(probs_all_data, indexes_all_data)\n",
    "\n",
    "# Compute soft accuracy\n",
    "soft_accuracy, soft_accuracy_dict = compute_soft_accuracies(true_data, predicted_data)\n",
    "\n",
    "print(\"Accuracy (acc/std):\", accuracy, \"\\nSoft accuracy (acc/std):\", \n",
    "      soft_accuracy, \"\\nPerplexity (perpl/std): \", perplexity)\n",
    "print(\"\\n\\nAccuracy dict:\", accuracy_dict, \"\\n\\nSoft accuracy dict:\", \n",
    "      soft_accuracy_dict, \"\\n\\nPerplexity dict: \", perplexity_dict)\n",
    "# print(true_data)\n",
    "# print(predicted_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "accur_dict = accuracy_dict\n",
    "soft_accur_dict = soft_accuracy_dict\n",
    "perplex_dict = perplexity_dict\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "# Filter out the empty values for perplexity\n",
    "x_values_perplexity = [x for x in accur_dict.keys() if perplex_dict[x]]\n",
    "perplexity_values = [perplex_dict[x] for x in x_values_perplexity]\n",
    "\n",
    "# Plot accuracy and soft accuracy on one graph\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "plt.plot(x_values_perplexity, perplexity_values, label='Perplexity', color='orange')\n",
    "\n",
    "# Set labels and title\n",
    "plt.xlabel('Ranges')\n",
    "plt.ylabel('Perplexity')\n",
    "plt.title('Perplexity Over sequence')\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.grid(True)\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# Create lists for x-axis and corresponding values for each dictionary\n",
    "# Filter out the empty values\n",
    "x_values = [x for x in accur_dict.keys() if accur_dict[x] and soft_accur_dict[x] and perplex_dict[x]]\n",
    "accuracy_values = [accur_dict[x] for x in x_values]\n",
    "soft_accuracy_values = [soft_accur_dict[x] for x in x_values]\n",
    "\n",
    "\n",
    "# Plot the data\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "plt.plot(x_values, accuracy_values, label='Accuracy')\n",
    "plt.plot(x_values, soft_accuracy_values, label='Soft Accuracy')\n",
    "\n",
    "# Set labels and title\n",
    "plt.xlabel('Range')\n",
    "plt.ylabel('Value')\n",
    "plt.title('Accuracy, Soft Accuracy Over ranges')\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.grid(True)\n",
    "plt.show()\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
